README.md
# flutter_residemenu

Residemenu for flutter,android and ios supported

## Screenshot

![ios](https://github.com/peng8350/flutter_residemenu/blob/master/arts/residemenu.gif)

## Usage
declare in pubspec.yaml

```

dependencies:
  ...

  residemenu:
    ^1.3.5

```


build() method to create my ResideMenu Widget,child is your contentView(if you use MaterialApp,mostly Scaffold),
about leftScafford,rightScafford,they are used to build your MenuView According to the layout of the public.
the more you can see the demo find how to use.If you do not meet your needs, you can use the ResideMenu.custom method.


```


       initState(){
          _menuController = MenuController(vsync:this,direction:ScrollDirection.Left);
       }

       dispose(){
          _menuController.dispose();
       }

      Widget buildItem(msg) {
        return new Material(
          color: Colors.transparent,
          child: new InkWell(
            child: const ResideMenuItem(
                title: "菜单一", icon: const Icon(Icons.home, color: Colors.grey),right: const Icon(Icons.arrow_forward,color:Colors.grey),),
            onTap: () {
              Scaffold
                  .of(context)
                  .showSnackBar(new SnackBar(content: new Text('你点击了$msg')));
            },
          ),
        );
      }

    new ResideMenu.scafford(

            decoration: new BoxDecoration(
                image: new DecorationImage(
                    image: new AssetImage("images/menu_background.png"),
                    fit: BoxFit.cover)),
            leftScafford: new MenuScaffold(
                                  header: new ConstrainedBox(

                                    constraints: new BoxConstraints(maxHeight: 80.0, maxWidth: 80.0),
                                    child: new CircleAvatar(

                                      backgroundImage: new AssetImage('images/author.jpeg'),
                                      radius: 40.0,
                                    ),
                                  ),
                                  children: <Widget>[
                                    buildItem("菜单一"),
                                    buildItem("菜单二"),
                                    buildItem("菜单三"),
                                    buildItem("菜单四"),
                                    buildItem("菜单五")
                                  ],
                                ),
            rightScafford: ...,
            controller: _menuController,
            child: ...
            )

```



## Table

| Attribute Name     |     Attribute Explain     | Parameter Type | Default Value  | requirement |
|---------|--------------------------|:-----:|:-----:|:-----:|
| child      | your content View   | Widget   |   null |  necessary |
| leftView,rightView | the Menu Content View     | Widget  | null | optional |
| elevation |   Content View shadow | double | 12.0 |optional |
| controller | Control menu behavior, get menu status, monitor menu open, close and other events.   | MenuController | null | optional |
| decoration | use to set bg and color in bottom   | BoxDecoration | null | optional |
| onOpen |   Event | (bool isOpenLeft) => Void | null |optional |
| onClose | Event   | () => Void | null | optional |
| onOffsetChange | when the child offset change it will callback(0~1)  | (double offset) => void | null | optional |
| enableFade | fadeColor cover content View   | bool | true | optional |
| enableScale | if you the gpu too high that lead the performance problem,it should be closed   | bool | true | optional |

## LICENSE

```
MIT License

Copyright (c) 2018 Jpeng

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
```

CHANGELOG.md
## [1.0.0] - 2018-5-10

* initalRelease

## [1.1.0] - 2018-5-11

* Add ScaleAlpha Effect when open or close

## [1.1.5] - 2018-5-18
* Fix the problem of state loss of the controller
* Add ResideMenuItem and Scfford to constructor Menu
* Provide another way to build MenuView

## [1.1.8] - 2018-5-21
* Fix my igore to limit scroll when Direction == LEFT or RIGHT

## [1.2.0] - 2018-5-22
* Fixed when content View clicks some gestureDector can not be closed.
* Fix itemExtent not used for Scafford

## [1.2.5] - 2018-5-30
* Add enableScale,enableFade option,ScaleAnimation has a proformance problem


## [1.3.1] - 2019-6-3
* Add enable3Drotate
* Remove bottom menuView and bg after closeMenu(improve performance)
* Optimizing internal code

## [1.3.2] - 2019-6-3
* Fix direction not working
* Move direction from ResideMenu to MenuController
* Fix onClose,onOpen callback

## [1.3.3] - 2019-6-4
* Fix decoration bug which invoke rebuild state

## [1.3.5] - 2019-6-5
* Fix init menuController not reference

## [1.3.6]
* enlarge start dx to trigger open remenu distance(if too small,user vertical gesture will open menu too)

## [1.3.7]
* interrupt close Modal when menu is open,when user press back
* fix residemenuitem  layout

lib/residemenu.dart
/*
 * Author: Jpeng
 * Email: peng8350@gmail.com
 * createTime: 2018-5-9 21:13
 */

library residemenu;

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';

typedef void OnOpen(bool isLeft);
typedef void OnClose();
typedef void OnOffsetChange(double offset);

enum ScrollDirection { LEFT, RIGHT, BOTH }
enum ScrollState { ScrollToLeft, NONE, ScrollToRight }

class ResideMenu extends StatefulWidget {
  // your content View
  final Widget child;

  //left or right Menu View
  final Widget leftView, rightView;

  //shadow elevation
  final double elevation;

  // it will control the menu Action,such as openMenu,closeMenu
  final MenuController controller;

  // used to set bottom bg and color
  final BoxDecoration decoration;

  final OnOpen onOpen;

  final OnClose onClose;

  final bool enableScale, enableFade, enable3dRotate;

  final OnOffsetChange onOffsetChange;

  ResideMenu.scaffold(
      {@required this.child,
      Widget leftBuilder,
      MenuScaffold leftScaffold,
      MenuScaffold rightScaffold,
      this.decoration: const BoxDecoration(),
      this.elevation: 12.0,
      this.onOpen,
      this.enableScale: true,
      this.enableFade: true,
      this.enable3dRotate: false,
      this.onClose,
      this.onOffsetChange,
      this.controller,
      Key key})
      : assert(child != null),
        leftView = leftScaffold,
        rightView = rightScaffold,
        super(key: key);

  ResideMenu.custom(
      {@required this.child,
      this.leftView,
      this.rightView,
      this.decoration: const BoxDecoration(color: const Color(0xff0000)),
      this.elevation: 12.0,
      this.onOpen,
      this.onClose,
      this.enable3dRotate: false,
      this.enableScale: true,
      this.enableFade: true,
      this.onOffsetChange,
      this.controller,
      Key key})
      : assert(child != null),
        super(key: key);

  @override
  _ResideMenuState createState() => new _ResideMenuState();
}

class _ResideMenuState extends State<ResideMenu>
    with SingleTickerProviderStateMixin {
  //determine width
  double _width = 0.0;
  MenuController _controller;
  ValueNotifier<ScrollState> _scrollState =
      ValueNotifier<ScrollState>(ScrollState.NONE);

  void _onScrollMove(DragUpdateDetails details) {
    double offset = details.delta.dx / _width * 2.0;
    if (_controller.value == 0.0) {
      if (details.delta.dy.abs() > details.delta.dx.abs() ||
          details.delta.dx.abs() < 10) return;
    }
    _controller.value += offset;
  }

  void _onScrollEnd(DragEndDetails details) {
    if (_controller.value > 0.5) {
      _controller.openMenu(true);
    } else if (_controller.value < -0.5) {
      _controller.openMenu(false);
    } else {
      _controller.closeMenu();
    }
  }

  void _handleScrollChange() {
    _scrollState.value = _controller.value == 0.0
        ? ScrollState.NONE
        : _controller.value > 0.0
            ? ScrollState.ScrollToLeft
            : ScrollState.ScrollToRight;
    if (widget.onOffsetChange != null) {
      widget.onOffsetChange(_controller.value.abs());
    }
  }

  void _handleScrollEnd(AnimationStatus status) {
    if (status == AnimationStatus.completed) {
      if (_controller.value == 1.0) {
        if (widget.onOpen != null) {
          widget.onOpen(true);
        }
      } else if (_controller.value == -1.0) {
        if (widget.onOpen != null) {
          widget.onOpen(false);
        }
      } else {
        if (widget.onClose != null) {
          widget.onClose();
        }
      }
    }
  }

  @override
  void didChangeDependencies() {
    // TODO: implement didChangeDependencies
    _update();
    super.didChangeDependencies();
  }

  // update listener
  void _update() {
    final MenuController newController = widget.controller ??
        MenuController(vsync: this, direction: ScrollDirection.LEFT);
    if (newController == null || newController == _controller) return;
    if (_controller != null)
      _controller
        ..removeListener(_handleScrollChange)
        ..removeStatusListener(_handleScrollEnd);
    _controller = newController;
    _controller
      ..addListener(_handleScrollChange)
      ..addStatusListener(_handleScrollEnd);
  }

  @override
  void initState() {
    // TODO: implement initState
    _scrollState.addListener(() {
      setState(() {});
    });
    super.initState();
  }

  @override
  void dispose() {
    // TODO: implement dispose
    _scrollState.dispose();
    _controller
      ..removeListener(_handleScrollChange)
      ..removeStatusListener(_handleScrollEnd);
    if (widget.controller == null) {
      _controller.dispose();
    }
    super.dispose();
  }

  @override
  void didUpdateWidget(ResideMenu oldWidget) {
    // TODO: implement didUpdateWidget
    _update();
    super.didUpdateWidget(oldWidget);
  }

  @override
  Widget build(BuildContext context) {
    return new LayoutBuilder(builder: (context, cons) {
      _width = cons.biggest.width;
      return WillPopScope(
        child: GestureDetector(
          onPanUpdate: _onScrollMove,
          onPanEnd: _onScrollEnd,
          child: new Stack(
            children: <Widget>[
              _scrollState.value != ScrollState.NONE
                  ? new Container(
                      decoration: widget.decoration,
                    )
                  : null,
              _scrollState.value != ScrollState.NONE
                  ? _MenuTransition(
                      offset: _controller,
                      child: new Container(
                          margin: new EdgeInsets.only(
                              left: (_scrollState.value ==
                                      ScrollState.ScrollToRight
                                  ? cons.biggest.width * 0.3
                                  : 0.0),
                              right: (_scrollState.value ==
                                      ScrollState.ScrollToLeft
                                  ? cons.biggest.width * 0.3
                                  : 0.0)),
                          child: _scrollState.value == ScrollState.ScrollToLeft
                              ? widget.leftView
                              : widget.rightView),
                    )
                  : null,
              _ContentTransition(
                  enableScale: widget.enableScale,
                  enable3D: widget.enable3dRotate,
                  child: new Stack(
                    children: <Widget>[
                      Container(
                        child: widget.child,
                        decoration: new BoxDecoration(boxShadow: <BoxShadow>[
                          new BoxShadow(
                            color: const Color(0xcc000000),
                            offset: const Offset(-2.0, 2.0),
                            blurRadius: widget.elevation * 0.66,
                          ),
                        ]),
                      ),
                      _scrollState.value != ScrollState.NONE
                          ? AnimatedBuilder(
                              animation: _controller,
                              builder: (c, w) {
                                return GestureDetector(
                                  child: Container(
                                    width: cons.biggest.width,
                                    height: cons.biggest.height,
                                    color: new Color.fromARGB(
                                        !widget.enableFade
                                            ? 0
                                            : (125 * _controller.value.abs())
                                                .toInt(),
                                        0,
                                        0,
                                        0),
                                  ),
                                  onTap: () {
                                    _controller.closeMenu();
                                  },
                                );
                              },
                            )
                          : null,
                    ].where((child) => child != null).toList(),
                  ),
                  menuOffset: _controller),
            ].where((child) => child != null).toList(),
          ),
        ),
        onWillPop: () async {
          if (_controller.value != 0) {
            _controller.closeMenu();
            return false;
          }
          return true;
        },
      );
    });
  }
}

class ResideMenuItem extends StatelessWidget {
  final String title;

  final TextStyle titleStyle;

  final Widget icon, right;

  final double height;

  final double midSpacing, leftSpacing, rightSpacing;

  const ResideMenuItem(
      {Key key,
      this.title: "菜单物品",
      this.titleStyle: const TextStyle(
          inherit: true, color: const Color(0xffdddddd), fontSize: 15.0),
      this.icon: const Text(""),
      this.right: const Icon(
        Icons.arrow_forward_ios,
        color: Colors.grey,
      ),
      this.height: 50.0,
      this.leftSpacing: 15.0,
      this.rightSpacing: 50.0,
      this.midSpacing: 30.0});

  @override
  Widget build(BuildContext context) {
    // TODO: implement build
    return new Container(
        height: 40.0,
        child: new Row(
          mainAxisAlignment: MainAxisAlignment.spaceBetween,
          children: <Widget>[
            Container(
              child: Row(
                children: <Widget>[
                  icon,
                  Container(
                    width: midSpacing,
                  ),
                  Text(
                    title,
                    style: titleStyle,
                  )
                ],
              ),
              margin: EdgeInsets.only(left: leftSpacing),
            ),
            Padding(
              child: right,
              padding: EdgeInsets.only(right: rightSpacing),
            )
          ],
        ));
  }
}

class _MenuTransition extends AnimatedWidget {
  final Widget child;

  final Animation<double> offset;

  _MenuTransition({@required this.child, this.offset, Key key})
      : super(key: key, listenable: offset);

  @override
  Widget build(BuildContext context) {
    // TODO: implement build
    return new LayoutBuilder(builder: (context, cons) {
      double width = cons.biggest.width;
      double height = cons.biggest.height;
      final Matrix4 transform = new Matrix4.identity()
        ..scale(2 - offset.value.abs(), 2 - offset.value.abs(), 1.0);
      return Opacity(
        opacity: offset.value.abs(),
        child: new Transform(
            transform: transform,
            child: child,
            origin: new Offset(width / 2, height / 2)),
      );
    });
  }
}

class _ContentTransition extends AnimatedWidget {
  final Widget child;

  final bool enableScale;
  final bool enable3D;

  _ContentTransition(
      {@required this.child,
      @required Animation<double> menuOffset,
      Key key,
      this.enable3D,
      this.enableScale})
      : super(key: key, listenable: menuOffset);

  Animation<double> get menuOffset => listenable;

  @override
  Widget build(BuildContext context) {
    // TODO: implement build
    return new LayoutBuilder(builder: (context, cons) {
      double width = cons.biggest.width;
      double val = menuOffset.value;
      final Matrix4 transform = new Matrix4.identity();
      if (enable3D) {
        transform.setEntry(3, 2, 0.0008);
        transform.rotateY(val * 0.8);
        transform.transposeRotation();
      }
      if (enableScale) {
        transform.scale(1.0 - 0.25 * val.abs(), 1 - 0.25 * val.abs(), 1.0);
      }

      transform.translate(val * 0.8 * width);
      return Transform(
        alignment: Alignment.center,
        transform: transform,
        child: child,
      );
    });
  }
}

class MenuController extends AnimationController {
  bool _isOpenLeft = false;
  bool _isOpenRight = false;

  MenuController(
      {TickerProvider vsync,
      ScrollDirection direction: ScrollDirection.LEFT,
      Duration openDuration: const Duration(milliseconds: 300)})
      : super(
            vsync: vsync,
            lowerBound: direction == ScrollDirection.LEFT ? 0.0 : -1.0,
            upperBound: direction == ScrollDirection.RIGHT ? 0.0 : 1.0,
            duration: openDuration,
            value: 0.0);

  Future<void> openMenu(bool left) {
    return animateTo(left ? 1.0 : -1.0).then((_) {
      _isOpenLeft = left;
    });
  }

  Future<void> closeMenu() {
    return animateTo(0.0).then((_) {
      _isOpenLeft = false;
      _isOpenRight = false;
    });
  }

  bool get isOpenLeft => _isOpenLeft;

  bool get isOpenRight => _isOpenLeft;

  bool get isClose => !_isOpenLeft && !_isOpenRight;
}

class MenuScaffold extends StatelessWidget {
  final List<Widget> children;
  final Widget header;
  final Widget footer;
  final double itemExtent;
  final double topMargin;

  MenuScaffold(
      {Key key,
      @required this.children,
      this.topMargin: 100.0,
      Widget header,
      Widget footer,
      this.itemExtent: 40.0})
      : assert(children != null),
        header = header ?? new Container(height: 20.0),
        footer = footer ?? new Container(height: 20.0),
        super(key: key);

  @override
  Widget build(BuildContext context) {
    // TODO: implement build
    return new Container(
      padding: new EdgeInsets.only(top: this.topMargin),
      child: new Column(
        crossAxisAlignment: CrossAxisAlignment.center,
        children: <Widget>[
          header,
          new ListView(
            physics: const NeverScrollableScrollPhysics(),
            itemExtent: this.itemExtent,
            shrinkWrap: true,
            children: children,
          ),
          footer
        ],
      ),
    );
  }
}


example/lib/main.dart
import 'MainActivity.dart';
import 'package:flutter/material.dart';

void main() => runApp(new MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return new MaterialApp(
      title: 'Residemenu',
      debugShowCheckedModeBanner: false,
      showPerformanceOverlay: true,
      theme: new ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or press Run > Flutter Hot Reload in IntelliJ). Notice that the
        // counter didn't reset back to zero; the application is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: new MyHomePage(title: 'Residemenu'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  _MyHomePageState createState() => new _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return new Scaffold(
      body: new MainActivity(),
    );
  }
}


example/lib/MainActivity.dart
import 'package:flutter/material.dart';
import 'package:residemenu/residemenu.dart';

class MainActivity extends StatefulWidget {
  @override
  State<StatefulWidget> createState() {
    // TODO: implement createState
    return new _MainActivityState();
  }
}

class _MainActivityState extends State<MainActivity>
    with TickerProviderStateMixin {
  MenuController _menuController;

  Widget buildItem(String msg1) {
    return new Material(
      color: Colors.transparent,
      child: new InkWell(
        child: ResideMenuItem(
            title: msg1, icon: const Icon(Icons.home, color: Colors.grey)),
        onTap: () {
          Scaffold.of(context)
              .showSnackBar(new SnackBar(content: new Text('你点击了$msg1')));
        },
      ),
    );
  }

  @override
  void dispose() {
    // TODO: implement dispose
    _menuController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return new ResideMenu.scaffold(
      enable3dRotate: true,
      controller: _menuController,
      decoration: new BoxDecoration(
          image: new DecorationImage(
              image: new AssetImage("images/menu_background.png"),
              fit: BoxFit.none)),
      leftScaffold: new MenuScaffold(
        header: new ConstrainedBox(
          constraints: new BoxConstraints(maxHeight: 80.0, maxWidth: 80.0),
          child: new CircleAvatar(
            backgroundImage: new AssetImage('images/author.jpeg'),
            radius: 40.0,
          ),
        ),
        children: <Widget>[
          buildItem("菜单一a"),
          buildItem("菜单二"),
          buildItem("菜单三"),
          buildItem("菜单四"),
          buildItem("菜单五")
        ],
      ),
      rightScaffold: new MenuScaffold(
        header: new CircleAvatar(
          backgroundImage: new AssetImage('images/author.jpeg'),
          radius: 40.0,
        ),
        children: <Widget>[
          buildItem("菜单一"),
          buildItem("菜单二"),
          buildItem("菜单三"),
          buildItem("菜单四"),
          buildItem("菜单五")
        ],
      ),
      child: new Scaffold(
        body: new Column(
          children: <Widget>[
            new Text('水平滚动组件测试'),
            new Container(
              height: 100.0,
              child: new ListView.builder(
                scrollDirection: Axis.horizontal,
                shrinkWrap: true,
                itemCount: 30,
                itemBuilder: (context, index) => new Image.asset(
                    "images/menu_background.png",
                    width: 400.0,
                    height: 100.0,
                    fit: BoxFit.cover),
              ),
            ),
            WillPopScope(
              child: Expanded(
                child: new ListView(
                  children: <Widget>[
                    new Text('sd'),
                    new Text('sd'),
                    new Text('sd'),
                    new Text('sd'),
                    new Text('sd'),
                    new Text('sd'),
                    new Text('sd'),
                    new Text('sd'),
                    new Text('sd'),
                    new Text('sd')
                  ],
                  itemExtent: 150.0,
                ),
              ),
              onWillPop: () async {
                return true;
              },
            )
          ],
        ),
        appBar: new AppBar(
          leading: new GestureDetector(
            child: const Icon(Icons.menu),
            onTap: () {
              _menuController.openMenu(true);
            },
          ),
          actions: <Widget>[
            new GestureDetector(
              child: const Icon(Icons.menu),
              onTap: () {
                _menuController.openMenu(false);
              },
            )
          ],
          // Here we take the value from the MyHomePage object that was created by
          // the App.build method, and use it to set our appbar title.
          title: new Text('ResideMenu'),
        ),
      ),
      onClose: () {
        print("closed");
      },
      onOpen: (left) {
        print(left);
      },
      onOffsetChange: (offset) {},
    );
  }

  @override
  void initState() {
    // TODO: implement initState
    super.initState();
    _menuController =
        new MenuController(vsync: this, direction: ScrollDirection.LEFT);
  }
}
